src/v/storage/tests/concat_segment_reader_test.cc

```diff
@@ -0,0 +1,165 @@
+#include "storage/log_manager.h"
+#include "storage/tests/utils/disk_log_builder.h"
+#include "test_utils/tmp_dir.h"
+
+#include <seastar/testing/thread_test_case.hh>
+#include <seastar/util/defer.hh>
+
+constexpr size_t segment_size{1024};
+
+using namespace storage;
+
+size_t copy_stream(concat_segment_reader_view& cv) {
+    auto concat_str = cv.take_stream();
+    iobuf buffer{};
+    auto ostream = make_iobuf_ref_output_stream(buffer);
+    ss::copy(concat_str, ostream).get();
+
+    // We do not actually need to close this, because the stream has been moved
+    // out of this view.
+    cv.close().get();
+
+    // Closes the stream, the data source and the
+    // concat_segment_data_source_impl
+    concat_str.close().get();
+    auto sz = buffer.size_bytes();
+    ostream.close().get();
+    return sz;
+}
+
+SEASTAR_THREAD_TEST_CASE(test_multiple_segments_read) {
+    temporary_dir tmp_dir("concat_segment_read");
+    auto data_path = tmp_dir.get_path();
+
+    disk_log_builder b{log_config{
+      log_config::storage_type::disk,
+      data_path.string(),
+      segment_size,
+      debug_sanitize_files::yes,
+    }};
+
+    b | start(ntp_config{{"test_ns", "test_tpc", 0}, {data_path}});
+    size_t start_offset = 0;
+    for (auto i = 0; i < 4; ++i) {
+        b | add_segment(start_offset) | add_random_batch(start_offset, 10);
+        auto& segment = b.get_segment(i);
+        start_offset = segment.offsets().dirty_offset + model::offset{1};
+    }
+
+    const auto& log_segments = b.get_log_segments();
+    auto segments = std::vector<ss::lw_shared_ptr<segment>>{
+      log_segments.begin(), log_segments.end()};
+
+    // Start 20 bytes into the first segment
+    size_t start_pos = 20;
+
+    // End 20 bytes before the last segment
+    size_t end_pos = log_segments.back()->file_size() - 20;
+
+    concat_segment_reader_view cv{
+      segments, start_pos, end_pos, ss::default_priority_class()};
+    BOOST_REQUIRE_EQUAL(
+      b.get_disk_log_impl().size_bytes() - 40, copy_stream(cv));
+    b.stop().get();
+}
+
+SEASTAR_THREAD_TEST_CASE(test_single_segment_read_with_bounds) {
+    temporary_dir tmp_dir("concat_segment_read");
+    auto data_path = tmp_dir.get_path();
+
+    using namespace storage;
+
+    disk_log_builder b{log_config{
+      log_config::storage_type::disk,
+      data_path.string(),
+      segment_size,
+      debug_sanitize_files::yes,
+    }};
+
+    b | start(ntp_config{{"test_ns", "test_tpc", 0}, {data_path}})
+      | add_segment(0) | add_random_batch(0, 10);
+
+    const auto& log_segments = b.get_log_segments();
+    auto segments = std::vector<ss::lw_shared_ptr<segment>>{
+      log_segments.begin(), log_segments.end()};
+
+    size_t start_pos = 20;
+    size_t end_pos = log_segments.back()->file_size() - 20;
+
+    concat_segment_reader_view cv{
+      segments, start_pos, end_pos, ss::default_priority_class()};
+    BOOST_REQUIRE_EQUAL(
+      b.get_disk_log_impl().size_bytes() - 40, copy_stream(cv));
```

Comment:
Would be nice to check that the content matches the expectation too.

Reply:
I planned to add these tests in the next PR which will need to test compacted segment uploads using the ntp archiver, but I think this test can be added in this PR as well, I will look into adding it to this PR.

Reply:
I don't mind. Up to you if you think a follow up PR is more suitable.

Reply:
Since there is some more work going on in this PR, I added a test `test_multiple_segments_read_with_content_verification` to do content verification using header batch types.